Use a monotonic clock for WebsocketSession's session_time

outputenable opened the pull request:
**What this PR solves**

Automatically scheduled actions like ping etc. and various other timeouts rely on WebsocketSession's session_time property to determine when to execute/fire.  This property is in turn based on Python's
time.time() which does not guarantee to always return non-decreasing values though.  To quote the module documentation,

> "[...] it can return a lower value than a previous call if the system clock has been set back between the two calls."  Source: https://docs.python.org/2/library/time.html#time.time

In order to not fall victim to such quirks it is common for timeouts and stuff like that to use a so-called monotonic clock which---as their name implies---cannot go backwards due to e.g. system clock updates.  Accordingly Python 3 also introduced such a clock in release 3.3, see:

- https://docs.python.org/3/library/time.html#time.monotonic

A rationale can be found in the associated "PEP 418 -- Add monotonic time, performance counter, and process time functions":

- https://www.python.org/dev/peps/pep-0418/

**How it is done**

There is an (actively maintained, by the looks of it) compatibility library available on PyPI fittingly called monotonic which provides a suitable monotonic clock for both Python 2 and 3:

- https://pypi.org/project/monotonic/

This PR suggests to replace the time.time() calls to initialize and update the session time with calls to monotonic.monotonic() which is not affected by changes of the system time.  This of course introduces a new runtime dependency on the monotonic module.  If import monotonic fails we can simply fall back on time.time() for backwards compatibility.

**What to look out for**

Event.receive_time is still using time.time() as before.  (Note that the reference point of the values returned by a monotonic clock is generally undefined.)

**Review**

- [x] Ready for review

outputenable committed:
Use a monotonic clock for WebsocketSession's session_time

This will try to use the facilities provided by module monotonic.  If
the import fails we'll fall back on using time.time() as before.

willmcgugan commented on lomond/session.py:
```diff
@@ -13,10 +13,14 @@
 
+try:
+    from monotonic import monotonic as monotonic_time
+except:
```
Best to explicitly catch an `ImportError` in the off chance the module has any import time side-effects.

willmcgugan commented on lomond/session.py:
```diff
@@ -13,10 +13,14 @@
+try:
+    from monotonic import monotonic as monotonic_time
+except:
+    from time import time as monotonic_time
```
If its imported this way, it's not strictly `monotonic_time`. Suggest calling it `get_time`

willmcgugan reviewed (changes requested):
First PR? Congrats.

Good call, but I'd make a couple of changes. Also could you add `monotonic` to requirements?

willmcgugan commented:
@outputenable Think there is a misconfiguration our side. Seems to work with PRs from within our organisation. Will look in to that.

outputenable committed:
Explicitly catch only ImportError when importing monotonic

outputenable committed:
Rename monotonic_time to get_time

outputenable commented:
Agreed, I've made the code changes you suggested, looks better now.  Could you expand a little on the requirements thing though?  Should I add monotonic in setup.py, or elsewhere?  (Oh and thank you for making such a nice library available as open source!)

willmcgugan commented:
Glad you find it useful!

If you could add `monotonic==1.5` to setup.py (install_requires), that will ensure monotonic is installed alongside Lomond. On reflection, you won't need to catch that ImportError, as it should always be installed.

I may actually end up copying monotonic.py in to this repos, if the licenses are compatible. But I'll leave that for another update.

Thanks.

outputenable committed:
Make monotonic a mandatory requirement and adjust import

outputenable commented:
Ah OK, I thought monotonic would be optional and wasn't sure where/how to add such a requirement.  So I've

- added install_requires `monotonic>=1.5` (>= instead of == as that's also used for the six dependency);
- removed the try/except around the import;
- renamed get_time to monotonic_time again, which IMO is appropriate now and avoids confusion.

What do you think?

willmcgugan commented:
Hi @outputenable I figured out the circleci issue. The env var is only available in our organisation. Could you please add `ignore_outcome=true` to tox.ini, under the [testenv:coverage] section? That will let us merge this.

outputenable committed:
Set ignore_outcome=true for tox testenv:coverage

This should fix the CircleCI issue.

outputenable committed:
Merge branch 'master' into monotonic-session-time

outputenable committed:
Fix duplicate name 'ignore_outcome' in tox.ini

outputenable commented:
Uh-oh, maybe I shouldn't have clicked that 'Merge' button so quickly...!?  \*sigh\*

Anyway, the Coveralls issue seems to be solved, but CircleCI hit a different problem now which to me looks like a genuine test failure of test_unresponsive.  I'm not familiar with tox or freezegun, but that test apparently relies on a fake system time and freeze_time() doesn't affect monotonic.monotonic().  There's a [freezegun pull request to support Python 3's time.monotonic()](https://github.com/spulec/freezegun/pull/199), but that's been dormant for quite a while and probably wouldn't help us anyway.  I'm afraid this is a little over my head still (Python is not my primary language) and I'll have to look into it some more in the next couple days.

willmcgugan commented:
I'll take a look at this soon. Suspect we can rework the tests to not require freezegun...

barfle commented:
Thanks for the original idea here. We have now incorporated this improvement (implemented fresh in-tree, not cherry-picked) in PR #93: https://github.com/wildfoundry/dataplicity-lomond/pull/93\n\nSpecifically, session timing now uses a monotonic clock path with compatibility fallback and test coverage.

barfle commented:
Closing as superseded by #93, where this idea was re-implemented and validated with current CI/test coverage.
